### PR TITLE
Lambda authorizer functions

### PR DESCRIPTION
**Pull request by alessio-creo**
needs #46

### CODE REVIEW

**Comment by micdes-pagopa** [comment]
`1733 files changed` 😬

File: src/main/assets/external_authorizer/lambda_authorizer.js
Comment (micdes-pagopa) [inline src/main/assets/external_authorizer/lambda_authorizer.js:1]: I can't see any invocations nor logs for this Lambda in the dev environment. Can you confirm this code has been tested?

File: src/main/assets/cognito_authorizer/lambda_authorizer.js
Comment (micdes-pagopa) [inline src/main/assets/cognito_authorizer/lambda_authorizer.js:1]: I can't see any logs for this function. Please check

File: src/main/assets/external_authorizer/lambda_authorizer.js
Comment (alessio-creo) [reply src/main/assets/external_authorizer/lambda_authorizer.js:1]: We've tested with cognito jwt. Ther's no permission for logging, I'm going to fix.
Then we'll tet again with the new jwt

File: src/main/assets/cognito_authorizer/lambda_authorizer.js
Comment (alessio-creo) [reply src/main/assets/cognito_authorizer/lambda_authorizer.js:1]: Tested. Works

**Review by micdes-pagopa — CHANGES_REQUESTED** [review]

File: src/main/assets/external_authorizer/lambda_authorizer.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    strictSsl: true,
+    jwksUri: process.env.JWKS_URI
+})
+
+const verificationOptions = {
+    "algorithms": "RS256"
+}
+
+
+function getSigningKey (header = decoded.header, callback) {
```
[the remark this answers is not in the document]
Comment (alessio-creo) [reply src/main/assets/external_authorizer/lambda_authorizer.js:14]: https://github.com/auth0/node-jsonwebtoken

Results of the library

File: src/main/assets/external_authorizer/lambda_authorizer.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    console.log("Getting payload")
+    var token = event.headers.Authorization.split(' ')[1];
+    
+    console.log("Generating authorization policy")
+
+    jwt.verify(token, getSigningKey, verificationOptions, function (error) {
+        if (error) {
+            callback(null, generatePolicy('Deny', event.methodArn));
+            console.log("User NOT allowed to perform the API call")
```
[the remark this answers is not in the document]
Comment (alessio-creo) [reply src/main/assets/external_authorizer/lambda_authorizer.js]: done

File: src/main/assets/external_authorizer/lambda_authorizer.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        } else {
+            callback(null, generatePolicy( 'Allow', event.methodArn));
+            console.log("User allowed to perform the API call")
```
[the remark this answers is not in the document]
Comment (alessio-creo) [reply src/main/assets/external_authorizer/lambda_authorizer.js]: done

File: src/main/assets/cognito_authorizer/lambda_authorizer.js
[the remark this answers is not in the document]
Comment (alessio-creo) [reply src/main/assets/cognito_authorizer/lambda_authorizer.js:1]: done

File: src/main/assets/external_authorizer/lambda_authorizer.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    cacheMaxAge: 86400000,
+    rateLimit: true,
+    jwksRequestsPerMinute: 10,
+    strictSsl: true,
+    jwksUri: process.env.JWKS_URI
+})
+
+const verificationOptions = {
+    "algorithms": "RS256"
+}
```
[the remark this answers is not in the document]
Comment (alessio-creo) [reply src/main/assets/external_authorizer/lambda_authorizer.js]: done

File: src/main/43-lambda-authorizer.tf [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  runtime          = "nodejs16.x"
+  environment {
+    variables = {
+      ENV          = var.env
+      ROLE_MAPPING = data.local_file.role_mapping.content
+    }
+  }
+}
+
+resource "aws_lambda_function" "external_authorizer" {
```
[the remark this answers is not in the document]
Comment (alessio-creo) [reply src/main/43-lambda-authorizer.tf:111]: done

File: src/main/assets/external_authorizer/lambda_authorizer.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
+const jwt = require('jsonwebtoken');
+const jwksClient = require('jwks-rsa');
+
+
+const keyClient = jwksClient({
+    cache: true,
+    cacheMaxAge: 86400000,
+    rateLimit: true,
+    jwksRequestsPerMinute: 10,
```
[the remark this answers is not in the document]
Comment (alessio-creo) [reply src/main/assets/external_authorizer/lambda_authorizer.js:9]: removed

**Review by micdes-pagopa — CHANGES_REQUESTED** [review]

File: src/main/43-lambda-authorizer.tf [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+resource "null_resource" "external_authorizer" {
+  provisioner "local-exec" {
+    command = "cd ${path.module}/assets/external_authorizer/ && npm install"
+  }
+
+  triggers = {
+    always_run = "${timestamp()}"
+  }
```
[the remark this answers is not in the document]
Comment (alessio-creo) [reply src/main/43-lambda-authorizer.tf:79]: it's okay

File: src/main/assets/external_authorizer/lambda_authorizer.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (alessio-creo) [reply src/main/assets/external_authorizer/lambda_authorizer.js:12]: https://www.npmjs.com/package/jsonwebtoken

Is the result of inner actions in jwt.verify method of the library

[the remark this answers is not in the document]
Comment (alessio-creo) [reply src/main/assets/external_authorizer/lambda_authorizer.js]: Removed the default value

File: src/main/43-lambda-authorizer.tf [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+  function_name    = "${var.app_name}-apigw-lambda-cognito-authorizer-${var.env}"
+  role             = aws_iam_role.lambda_authorizer_execution_role.arn
+  handler          = "lambda_authorizer.handler"
+  source_code_hash = data.archive_file.cognito_authorizer.output_base64sha256
+  runtime          = "nodejs16.x"
+  timeout          = 15
+  environment {
+    variables = {
+      ENV           = var.env
+      ROLE_MAPPING  = data.local_file.role_mapping.content
```
[the remark this answers is not in the document]
Comment (alessio-creo) [reply src/main/43-lambda-authorizer.tf]: Done

File: src/main/42-apigw.tf [hunk 9]
```diff
@@ -5,7 +5,7 @@ locals {
 resource "aws_api_gateway_rest_api" "apigw" {
   name        = "${var.app_name}-apigw-${var.env}"
   description = "${var.app_name} API Gateway"
-  body        = file(local.openapi_spec)
+  body        = templatefile(local.openapi_spec, { external_authorizer_arn = var.external_authorizer_uri, cognito_authorizer_arn = var.cognito_authorizer_uri })
```
Comment (micdes-pagopa) [inline src/main/42-apigw.tf:8]: Why pass these values as variables? You can get (or build) them from TF resources

Comment (alessio-creo) [reply src/main/42-apigw.tf:8]: The authorizer object is defined and created in the openapi specification

Comment (micdes-pagopa) [reply src/main/42-apigw.tf:8]: Those variables contain arns that refer to lambda functions that we create in TF, can't we build the arns using the functions' attributes?

Comment (alessio-creo) [reply src/main/42-apigw.tf]: Done

**Pull request merged**